sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala

```diff
@@ -3394,15 +3395,25 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
     }
   }
 
-  test("SPARK-30870: Column pruning shouldn't alias a nested column if it means the whole " +
-    "structure") {
-    val df = sql(
-      """
-        |SELECT explodedvalue.field
-        |FROM VALUES array(named_struct('field', named_struct('a', 1, 'b', 2))) AS (value)
-        |LATERAL VIEW explode(value) AS explodedvalue
-      """.stripMargin)
-    checkAnswer(df, Row(Row(1, 2)) :: Nil)
+  test("SPARK-30870: Column pruning shouldn't alias a nested column for the whole structure") {
+    withTable("t") {
+      val df = sql(
+        """
+          |SELECT value
+          |FROM VALUES array(named_struct('field', named_struct('a', 1, 'b', 2))) AS (value)
+        """.stripMargin)
+      df.write.format("parquet").saveAsTable("t")
```

Comment:
The below test coverage looks better. Last time, it checked only the exception occurs or not. BTW, do we need to store as table to avoid some local table optimization?

Reply:
Yea, if not store as a table, the query becomes a simple LocalRelation.

Reply:
Sorry if I am being dumb here but I can't fully follow why LocalRelation matters here. How does the current fix relate to the test fix here?

If we concern about `ConvertToLocalRelation`, it seems not effective here:

```

== Parsed Logical Plan ==
'Project ['explodedvalue.field]
+- 'Generate 'explode('value), false, as, ['explodedvalue]
   +- 'SubqueryAlias AS
      +- 'UnresolvedInlineTable [value], [List('array('named_struct(field, 'named_struct(a, 1, b, 2))))]

== Analyzed Logical Plan ==
field: struct<a:int,b:int>
Project [explodedvalue#219.field AS field#220]
+- Generate explode(value#218), false, as, [explodedvalue#219]
   +- SubqueryAlias AS
      +- LocalRelation [value#218]

== Optimized Logical Plan ==
Project [explodedvalue#219.field AS field#220]
+- Generate explode(value#218), [0], false, as, [explodedvalue#219]
   +- LocalRelation [value#218]

== Physical Plan ==
*(1) Project [explodedvalue#219.field AS field#220]
+- Generate explode(value#218), false, [explodedvalue#219]
   +- LocalTableScan [value#218]
```

Reply:
I changed the test added by SPARK-30870, because the root cause of the failure is not due to selecting whole structure. Here I change this test to more fit the original purpose of SPARK-30870: when selecting whole structure, column pruning shouldn't alias the nested column.

Reply:
The test added by SPARK-30870 is moved to `SchemaPruningSuite`.

Reply:
Both fixes fix the concrete SQL in SPARK-30870 independently because 
- the whole structure was selected (my pr fixes it) and 
- a nested column was selected from the output of a generate (this pr fixes it).

But this PR is better for cases that contain generate as it can handle nested field of a nested field https://github.com/apache/spark/pull/27702/files#diff-957112380b0a2ef014abc8227d0b70acR313-R318